internal/provider/model_fmc_access_control_policy.go

```diff
@@ -174,9 +451,222 @@ func (data *AccessControlPolicy) updateFromBody(ctx context.Context, res gjson.R
 	} else if data.DefaultActionSendSyslog.ValueBool() != false {
 		data.DefaultActionSendSyslog = types.BoolNull()
 	}
+
+	resLen := len(res.Get("dummy_categories").Array())
+	tflog.Debug(ctx, fmt.Sprintf("categories array: tf len %d, body len %d", len(data.Categories), resLen))
+	for i := len(data.Categories); i < resLen; i++ {
+		data.Categories = append(data.Categories, AccessControlPolicyCategories{})
+	}
+	if len(data.Categories) > resLen {
+		data.Categories = data.Categories[:resLen]
+	}
+
+	for i := range data.Categories {
+		r := res.Get(fmt.Sprintf("dummy_categories.%d", i))
```

Comment:
I'm thinking about adding this part to the gen/generator, under attribute `ordered_list: true` maybe (actually any `type: List` should use this logic, but the convention seems to be to use `List` instead of `Set`, somehow). For some JSON arrays we expect the API to change order, but there are JSON arrays where both API and TF know that order is crucial. @danischm

Reply:
If we automate the "ordered_list" fragments, we can probably auto-generate the entire `updateFromBody` here.

Reply:
I am not sure I fully understand what you are doing here, but I agree to adding it to the generator, so we can fully generate the (un)marshaling code. :-)

We can discuss the reasoning behind when to use Lists or Sets separately.

Reply:
Yeah, I'm seeing 1 very annoying problem with Set, which seems like Terraform's fault for now. Because of that problem I don't want to push for more usage of Set versus List.

I've added `ordered_list`, pushed it now.

The policy rules cannot be re-ordered: if we save to FMC rules {BLOCK, ALLOW} and later read back {ALLOW, BLOCK}, then the meaning of the entire policy is changed. We cannot disregard order when reading back from FMC. That's what the code is doing for `ordered_list`.